RFC for project goals

nikomatsakis opened the pull request:
This RFC presents the Rust project goal slate for 2024H2. The slate consists of 24 total project goals of which we have selected 3 as our "flagship goals":

* Release the Rust 2024 edition (owner: @traviscross)
* Bringing the Async Rust experience closer to parity with sync Rust (owners: @tmandry, @nikomatsakis)
* Resolving the biggest blockers to Linux building on stable Rust (owners: @joshtriplett, @nikomatsakis)

Flagship goals represent the goals expected to have the broadest overall impact.

**[This RFC is now in final comment period](https://github.com/rust-lang/rfcs/pull/3672#issuecomment-2236745798).** We are following an [unusual ratification procedure](https://rust-lang.zulipchat.com/#narrow/stream/435869-project-goals-2024h2/topic/Procedural.20next.20steps.20and.20timeline). Team leads are asked to review the list of asks for their team (see the reference-level-explanation) and confirm that their team is aligned. Now that all team leads have signed off, the RFC has entered FCP.

More details at https://rust-lang.github.io/rust-project-goals/.

[Rendered](https://github.com/nikomatsakis/rfcs/blob/project-goals-2024h2/text/3672-Project-Goals-2024h2.md)

lcnr commented on text/0000-Project-Goals-2024h2.md:
```diff
@@ -0,0 +1,384 @@
+
+## Motivation
+
+This RFC marks the first goal slate proposed under the experimental new roadmap process described in [RFC #3614](https://github.com/rust-lang/rfcs/pull/3614). It consists of NN project goals, of which we have selected three as **flagship goals**. Flagship goals represent the goals expected to have the broadest overall impact. 
```
```suggestion
This RFC marks the first goal slate proposed under the experimental new roadmap process described in [RFC #3614](https://github.com/rust-lang/rfcs/pull/3614). It consists of 24 project goals, of which we have selected three as **flagship goals**. Flagship goals represent the goals expected to have the broadest overall impact. 
```

nikomatsakis commented:
Hi everyone! 

This RFC follows an [unusual ratification procedure](https://rust-lang.zulipchat.com/#narrow/stream/435869-project-goals-2024h2/topic/Procedural.20next.20steps.20and.20timeline). Team leads are asked to review the list of asks for their team (see the reference-level-explanation) and confirm that their team is aligned. Leads should feel free to consult with team members and to raise concerns on their behalf. Once all team leads have signed off, the RFC will enter FCP.

Note that your name may appear more than once!

## cargo

* [ ] @ehuss (required, lead)
* [ ] @arlosi (optional)
* [ ] @Eh2406 (optional)
* [ ] @epage (optional)
* [ ] @joshtriplett (optional)
* [ ] @Muscraft (optional)
* [ ] @weihanglo (optional)

## clippy

* [ ] @flip1995 (required, lead)
* [ ] @Manishearth (required, lead)
* [ ] @Alexendoo (optional)
* [ ] @blyxyas (optional)
* [ ] @Centri3 (optional)
* [ ] @dswij (optional)
* [ ] @giraffate (optional)
* [ ] @Jarcho (optional)
* [ ] @llogiq (optional)
* [ ] @matthiaskrgr (optional)
* [ ] @xFrednet (optional)
* [ ] @y21 (optional)

## compiler

* [ ] @davidtwco (required, lead)
* [ ] @wesleywiser (required, lead)
* [ ] @Aaron1011 (optional)
* [ ] @cjgillot (optional)
* [ ] @compiler-errors (optional)
* [ ] @eddyb (optional)
* [ ] @estebank (optional)
* [ ] @jackh726 (optional)
* [ ] @lcnr (optional)
* [ ] @matthewjasper (optional)
* [ ] @michaelwoerister (optional)
* [ ] @nagisa (optional)
* [ ] @oli-obk (optional)
* [ ] @petrochenkov (optional)
* [ ] @pnkfelix (optional)

## crates-io

* [ ] @jtgeibel (required, lead)
* [ ] @Turbo87 (required, lead)
* [ ] @carols10cents (optional)
* [ ] @eth3lbert (optional)
* [ ] @hi-rustin (optional)
* [ ] @JohnTitor (optional)
* [ ] @LawnGnome (optional)
* [ ] @mdtro (optional)

## infra

* [ ] @jdno (required, lead)
* [ ] @shepmaster (required, lead)
* [ ] @kennytm (optional)
* [ ] @Kobzol (optional)
* [ ] @Mark-Simulacrum (optional)
* [ ] @pietroalbini (optional)

## lang

* [ ] @nikomatsakis (required, lead)
* [ ] @tmandry (required, lead)
* [ ] @joshtriplett (optional)
* [ ] @pnkfelix (optional)
* [ ] @scottmcm (optional)
* [ ] @traviscross (optional)

## leadership-council

* [ ] @eholk (optional)
* [ ] @ehuss (optional)
* [ ] @jackh726 (optional)
* [ ] @jamesmunns (optional)
* [ ] @m-ou-se (optional)
* [ ] @Mark-Simulacrum (optional)
* [ ] @technetos (optional)

## libs

* [ ] @Amanieu (required, lead)
* [ ] @m-ou-se (required, lead)
* [ ] @cuviper (optional)
* [ ] @joshtriplett (optional)
* [ ] @the8472 (optional)
* [ ] @thomcc (optional)

## libs-api

* [ ] @Amanieu (optional)
* [ ] @BurntSushi (optional)
* [ ] @dtolnay (optional)
* [ ] @joshtriplett (optional)
* [ ] @m-ou-se (optional)

## rust-analyzer

* [ ] @Veykril (required, lead)
* [ ] @flodiebold (optional)
* [ ] @HKalbasi (optional)
* [ ] @lnicola (optional)
* [ ] @lowr (optional)

## rustdoc

* [ ] @GuillaumeGomez (required, lead)
* [ ] @aDotInTheVoid (optional)
* [ ] @camelid (optional)
* [ ] @fmease (optional)
* [ ] @jsha (optional)
* [ ] @Manishearth (optional)
* [ ] @Nemo157 (optional)
* [ ] @notriddle (optional)

## rustdoc-frontend

* [ ] @camelid (optional)
* [ ] @GuillaumeGomez (optional)
* [ ] @jsha (optional)
* [ ] @notriddle (optional)

## types

* [ ] @jackh726 (required, lead)
* [ ] @lcnr (required, lead)
* [ ] @aliemjay (optional)
* [ ] @BoxyUwU (optional)
* [ ] @compiler-errors (optional)
* [ ] @nikomatsakis (optional)
* [ ] @oli-obk (optional)
* [ ] @spastorino (optional)

Nilstrieb commented:
many people didn't get pinged, you probably hit some anti spam protection :3

nikomatsakis commented:
## ~~cargo~~ 

* [x] @ehuss (**required**, lead)
* [ ] @arlosi (optional)
* [ ] @Eh2406 (optional)
* [ ] @epage (optional)
* [ ] @joshtriplett (optional)
* [ ] @Muscraft (optional)
* [ ] @weihanglo (optional)

nikomatsakis commented:
## clippy

* [ ] @flip1995 (**required**, lead)
* [ ] @Manishearth (**required**, lead)
* [ ] @Alexendoo (optional)
* [ ] @blyxyas (optional)
* [ ] @Centri3 (optional)
* [ ] @dswij (optional)
* [ ] @giraffate (optional)
* [ ] @Jarcho (optional)
* [ ] @llogiq (optional)
* [ ] @matthiaskrgr (optional)
* [ ] @xFrednet (optional)
* [ ] @y21 (optional)

nikomatsakis commented:
## compiler

* [x] @davidtwco (**required**, lead)
* [ ] @wesleywiser (**required**, lead)
* [ ] @Aaron1011 (optional)
* [ ] @cjgillot (optional)
* [x] @compiler-errors (optional)
* [ ] @eddyb (optional)
* [ ] @estebank (optional)
* [ ] @jackh726 (optional)
* [ ] @lcnr (optional)
* [ ] @matthewjasper (optional)
* [ ] @michaelwoerister (optional)
* [ ] @nagisa (optional)
* [ ] @oli-obk (optional)
* [ ] @petrochenkov (optional)
* [ ] @pnkfelix (optional)

nikomatsakis commented:
## crates-io

* [ ] @jtgeibel (**required**, lead)
* [ ] @Turbo87 (**required**, lead)
* [ ] @carols10cents (optional)
* [ ] @eth3lbert (optional)
* [ ] @hi-rustin (optional)
* [ ] @JohnTitor (optional)
* [ ] @LawnGnome (optional)
* [ ] @mdtro (optional)

nikomatsakis commented:
## edition

* [ ] @traviscross (**required**, lead)
* [x] @ehuss (**required**, lead)
* [ ] @m-ou-se (optional)

nikomatsakis commented:
## infra

* [ ] @jdno (**required**, lead)
* [ ] @shepmaster (**required**, lead)
* [ ] @kennytm (optional)
* [ ] @Kobzol (optional)
* [ ] @Mark-Simulacrum (optional)
* [ ] @pietroalbini (optional)

nikomatsakis commented:
## lang

* [x] @nikomatsakis (**required**, lead)
* [ ] @tmandry (**required**, lead)
* [ ] @joshtriplett (optional)
* [ ] @pnkfelix (optional)
* [ ] @scottmcm (optional)
* [ ] @traviscross (optional)

nikomatsakis commented:
## leadership-council

* [ ] @eholk (optional)
* [x] @ehuss (optional)
* [ ] @jackh726 (optional)
* [ ] @jamesmunns (optional)
* [ ] @m-ou-se (optional)
* [ ] @Mark-Simulacrum (optional)
* [ ] @technetos (optional)

nikomatsakis commented:
## libs

* [ ] @Amanieu (**required**, lead)
* [ ] @m-ou-se (**required**, lead)
* [ ] @cuviper (optional)
* [ ] @joshtriplett (optional)
* [ ] @the8472 (optional)
* [ ] @thomcc (optional)

nikomatsakis commented:
## libs-api

* [ ] @Amanieu (optional)
* [ ] @BurntSushi (optional)
* [ ] @dtolnay (optional)
* [ ] @joshtriplett (optional)
* [ ] @m-ou-se (optional)

nikomatsakis commented:
## rust-analyzer

* [ ] @Veykril (**required**, lead)
* [ ] @flodiebold (optional)
* [ ] @HKalbasi (optional)
* [ ] @lnicola (optional)
* [ ] @lowr (optional)

nikomatsakis commented:
## rustdoc

* [x] @GuillaumeGomez (**required**, lead)
* [ ] @aDotInTheVoid (optional)
* [ ] @camelid (optional)
* [ ] @fmease (optional)
* [ ] @jsha (optional)
* [ ] @Manishearth (optional)
* [ ] @Nemo157 (optional)
* [ ] @notriddle (optional)

nikomatsakis commented:
## rustdoc-frontend

* [ ] @camelid (optional)
* [ ] @GuillaumeGomez (optional)
* [ ] @jsha (optional)
* [ ] @notriddle (optional)

nikomatsakis commented:
## types

* [ ] @jackh726 (**required**, lead)
* [ ] @lcnr (**required**, lead)
* [ ] @aliemjay (optional)
* [ ] @BoxyUwU (optional)
* [x] @compiler-errors (optional)
* [x] @nikomatsakis (optional)
* [ ] @oli-obk (optional)
* [ ] @spastorino (optional)

athre0z commented on text/0000-Project-Goals-2024h2.md:
```diff
@@ -0,0 +1,521 @@
+    * never type fallback ([#123748](https://github.com/rust-lang/rust/issue/123748))
+    * and a [number of other potential changes](https://github.com/rust-lang/rust/issues?q=label%3AC-tracking-issue+label%3AA-edition-2024+label%3AS-tracking-ready-to-stabilize%2CS-tracking-needs-documentation+-label%3AS-tracking-impl-incomplete%2CS-tracking-design-concerns) that may be included if they make enough progress
+* [**Bringing the Async Rust experience closer to parity with sync Rust**](https://rust-lang.github.io/rust-project-goals/2024h2/./async.html) via:
+    * resolving the "send bound problem", thus enabling foundational, generic traits like Tower's [`Service`]() trait;
```
Missing link here, was probably supposed to point to tower doc?

```suggestion
    * resolving the "send bound problem", thus enabling foundational, generic traits like Tower's [`Service`](https://docs.rs/tower-service/latest/tower_service/trait.Service.html) trait;
```

nikomatsakis replied:
Thanks. I fixed it on the rust-project-goals repo.

flip1995 commented:
## clippy

We'd like to get the PRs https://github.com/rust-lang/rust-project-goals/pull/46 and https://github.com/rust-lang/rust-project-goals/pull/47 merged, if that is still possible. Other than that :heavy_check_mark: from my side.

jdno commented on text/0000-Project-Goals-2024h2.md:
```diff
@@ -0,0 +1,521 @@
+| [Next-generation trait solver](https://rust-lang.github.io/rust-project-goals/2024h2/next-solver.html)                                                            | [lcnr][]                | [rust-analyzer], [types]                                     |
+| [Assemble project goal slate](https://rust-lang.github.io/rust-project-goals/2024h2/Project-goal-slate.html)                                                      | [Niko Matsakis][]        | [leadership-council]                                         |
+| [Optimizing Clippy & linting](https://rust-lang.github.io/rust-project-goals/2024h2/optimize-clippy.html)                                                         | [Alejandra González][]             | [clippy]                                                     |
+| [Make Rustdoc Search easier to learn](https://rust-lang.github.io/rust-project-goals/2024h2/rustdoc-search.html)                                                  | [Michael Howell][]           | [infra], [leadership-council], [rustdoc], [rustdoc-frontend] |
```
It's not clear to me why this goal needs the infra-team. Reading through the proposal, it seems that the rustdoc and/or docs.rs teams are in a much better position to provide feedback.

nikomatsakis replied:
cc @GuillaumeGomez

nikomatsakis replied:
oops, sorry, this is @notriddle

nikomatsakis replied:
I guess the goal says "Collecting popular queries for review" -- I'll redirect to @rust-lang/docs-rs, I guess? I don't know how we would gather popular queries from rustdoc otherwise, unless you plan to have it "phone home".

nikomatsakis replied:
(You could of course use the ["telemetry without telemetry"](https://estebank.github.io/rustc-metrics.html) scheme described by @estebank, but I don't think that's what you were proposing.)

nikomatsakis replied:
Alternatively I could just remove this ask and we can figure it out as we go.

jdno replied:
Ah, I see! Thanks for the context!

For `docs.rs`, the docs.rs team should be the best point of contact. But `doc.rust-lang.org` is served as a static website from S3 and we're not collecting any logs today. So that might actually require the infra-team to help out after all.

syphar replied:
I somehow missed the mention above, sorry about the delay in responding. 

Docs.rs doesn't change rustdoc search in any way, so it's still only in the frontend. The only thing in our webserver (or CDN) logs would be fetching the search index file, not the search term itself. 

The only time when you would see the term in the access logs is probably when someone searches a term, and then manually refreshes the page, which feels like much less than 80% gains. 

So the only way to get some statistics on terms would be to add frontend JS code that hooks into the search itself on the frontend side, which probably would trigger the mentioned consent popup above?

nikomatsakis replied:
My proposal is that we remove this part of the project goal but that @notriddle you continue discussing with @syphar or @jdno to see if there's some easy step we can take here.

nikomatsakis replied:
Resolving since https://github.com/rust-lang/rust-project-goals/pull/51 removes this point, thanks @notriddle!

nikomatsakis commented:
> ## clippy
> 
> We'd like to get the PRs [rust-lang/rust-project-goals#46](https://github.com/rust-lang/rust-project-goals/pull/46) and [rust-lang/rust-project-goals#47](https://github.com/rust-lang/rust-project-goals/pull/47) merged, if that is still possible. Other than that ✔️ from my side.

OK, let me update that.

nikomatsakis commented:
## docs-rs

* [ ] @syphar (required, lead)
* [ ] @Byron (optional)
* [ ] @GuillaumeGomez (optional)
* [ ] @jsha (optional)
* [ ] @Nemo157 (optional)

I'm adding this team -- see [this thread](https://github.com/rust-lang/rfcs/pull/3672#discussion_r1673835686) for context.

jackh726 commented on text/0000-Project-Goals-2024h2.md:
```diff
@@ -0,0 +1,522 @@
+| [Expose experimental LLVM features for automatic differentiation and GPU offloading](https://rust-lang.github.io/rust-project-goals/2024h2/Rust-for-SciComp.html) | [Manuel Drehwald][]              | [compiler], [lang]                                             |
+| [Extend pubgrub to match cargo's dependency resolution](https://rust-lang.github.io/rust-project-goals/2024h2/pubgrub-in-cargo.html)                              | [Jacob Finkelman][]              | [cargo]                                                        |
+| [Implement "merged doctests" to save doctest time](https://rust-lang.github.io/rust-project-goals/2024h2/merged-doctests.html)                                    | [Guillaume Gomez][]      | [rustdoc]                                                      |
+| [Make Rustdoc Search easier to learn](https://rust-lang.github.io/rust-project-goals/2024h2/rustdoc-search.html)                                                  | [Michael Howell][]           | [docs-rs], [leadership-council], [rustdoc], [rustdoc-frontend] |
```
imo, I'm not sure I agree that the council should be a team here - reading the details, it seems like this is because of a blog post? But, I don't think this needs to go through the council.

nikomatsakis replied:
Removed.

jackh726 commented on text/0000-Project-Goals-2024h2.md:
```diff
@@ -0,0 +1,522 @@
+| [Stabilize doc_cfg](https://rust-lang.github.io/rust-project-goals/2024h2/doc_cfg.html)                                                                           | [Guillaume Gomez][]      | [rustdoc]                                                      |
+| [Stabilize parallel front end](https://rust-lang.github.io/rust-project-goals/2024h2/parallel-front-end.html)                                                     | [Sparrow Li][]          | [compiler]                                                     |
+| [Testing infra + contributors for a-mir-formality](https://rust-lang.github.io/rust-project-goals/2024h2/a-mir-formality.html)                                    | [Niko Matsakis][]        | [types]                                                        |
+| [Use annotate-snippets for rustc diagnostic output](https://rust-lang.github.io/rust-project-goals/2024h2/annotate-snippets.html)                                 | [Esteban Kuber][], [Scott Schafer][] | [compiler], [leadership-council]                               |
```
same here

jackh726 commented on text/0000-Project-Goals-2024h2.md:
```diff
@@ -0,0 +1,522 @@
+| ↳ [Rust 2024 Edition](https://rust-lang.github.io/rust-project-goals/2024h2/Rust-2024-Edition.html#ownership-and-team-asks)                                   | [TC][]         |                                                                                           |
+| *Top-level Rust blog post announcing result*                                                          |                      |                                                                                           |
+| ↳ [Assemble project goal slate](https://rust-lang.github.io/rust-project-goals/2024h2/Project-goal-slate.html#ownership-and-team-asks)                        | [Niko Matsakis][]        |                                                                                           |
+| *Top-level Rust blog post inviting feedback*                                                          |                      |                                                                                           |
+| ↳ [Make Rustdoc Search easier to learn](https://rust-lang.github.io/rust-project-goals/2024h2/rustdoc-search.html#ownership-and-team-asks)                    | [Michael Howell][]           |                                                                                           |
+| ↳ [Use annotate-snippets for rustc diagnostic output](https://rust-lang.github.io/rust-project-goals/2024h2/annotate-snippets.html#ownership-and-team-asks)   | [Esteban Kuber][], [Scott Schafer][] |                                                                                           |
```
As above, I don't think these are really under the council. At one point in time, core did approve posts to the main blog, but I think it's been a while since that requirement (the council surely hasn't been approving any).

jackh726 reviewed (approved):
This is really fantastic. Thank you @nikomatsakis for the work you've done to assemble this.

nikomatsakis commented:
@jackh726 when it comes to blog posts, I wasn't sure how to think about it. My general feeling is:

* Anybody can post on Inside Rust with zero overhead, as long as some team member +1's it, basically
* The main Rust blog has a higher bar, but I don't think we should keep it very high. I'd be happy to say "it's assumed that anything tied to a project goal can have blog posts on any Rust blog".

lcnr commented on text/0000-Project-Goals-2024h2.md:
```diff
@@ -0,0 +1,522 @@
+| Goal                                                                                                | Owner            | Notes              |
+| ---                                                                                                 | ---              | ---                |
+| *Design meeting*                                                                                    |                  |                    |
+| ↳ [Experiment with relaxing the Orphan Rule](https://rust-lang.github.io/rust-project-goals/2024h2/Relaxing-the-Orphan-Rule.html#ownership-and-team-asks)   | ![Help wanted][] | Up to 1, if needed |
```
This is the one project goal I feel somewhat uncomfortable with. I feel like I (and the wider types team) are already at capacity with the existing project goals and any additional bugs, unsoundnesses, or nominations during the next 6 months.

I feel especially uncertain as the required team input is very dependent on the owner and the direction they take this in. I don't trust that a single design meeting is sufficient, given that it will also require review work and asynchronous discussion. 

Even if we do not accept it, people can still work on it. The teams should be able to provide some feedback even for non-project goals. Having it be a project goal does worry me as I expect it to increase the pressure on team members to quickly respond to questions and proposals here.

nikomatsakis replied:
This is a solid point. I agree that the number of goals grew towards the end (initially there were few) and that we've got plenty to do. When thinking about this goal I didn't think properly about the *types* team interaction. This is likely a poor fit for an orphaned goal in particular as that implies we are looking to find owners for the external community, which probably works best when non-specialized knowledge is required.

nikomatsakis commented on text/0000-Project-Goals-2024h2.md:
```diff
@@ -0,0 +1,522 @@
+| ↳ [Rust 2024 Edition](https://rust-lang.github.io/rust-project-goals/2024h2/Rust-2024-Edition.html#ownership-and-team-asks)                                 | [TC][]     |                    |
+| ↳ [Associated type position impl trait](https://rust-lang.github.io/rust-project-goals/2024h2/ATPIT.html#ownership-and-team-asks)                           | [Oliver Scherer][]         |                    |
+| *Standard reviews*                                                                                  |                  |                    |
+| ↳ [Stabilize coherence coherence support](https://rust-lang.github.io/rust-project-goals/2024h2/next-solver.html#ownership-and-team-asks)                   | [lcnr][]            |                    |
```
This table is auto-generated from the goal file. The algorithm is a bit heuristic-y so I'll look at modifying the goal text.

nikomatsakis commented on text/0000-Project-Goals-2024h2.md:
```diff
@@ -0,0 +1,384 @@
+
+* **Rust must deliver on its promise of peak performance and high reliability.** Rust’s maximum advantage is in applications that require peak performance or low-level systems capabilities. We must continue to innovate and support those areas above all.
+* **Rust's goals require high productivity and ergonomics.** Being attentive to ergonomics broadens Rust impact by making it more appealing for projects that value reliability and maintenance but which don't have strict performance requirements.
+* **Slow and steady wins the race.** For this first round of goals, we want a small set that can be completed without undue stress. As the Rust open source org continues to grow, the set of goals can grow in size.
```
Yeah... it refers more to the flagship goals I think. :) Though also the set of goals grew quite a bit towards the end!

tmandry commented on text/0000-Project-Goals-2024h2.md:
```diff
@@ -0,0 +1,522 @@
+| ↳ [Trait for async iteration](https://rust-lang.github.io/rust-project-goals/2024h2/async.html#ownership-and-team-asks)                                                                       | [Eric Holk][]           |                    |
+| *Stabilization*                                                                                                                       |                  |                    |
+| ↳ ["Send bound" problem](https://rust-lang.github.io/rust-project-goals/2024h2/async.html#ownership-and-team-asks)                                                                            | [Niko Matsakis][]    |                    |
+| *Stabilization decision*                                                                                                              |                  |                    |
```
This list feels a bit long to me in terms of things we would commit to reviewing (going with the "start small" axiom).

I think the obvious candidate for removal is the one we haven't RFC'd yet and isn't part of a flagship goal, patterns of empty types (sorry @Nadrieril, feel free to make the case for it here). Of course it would still be good to consider it while it's fresh, if we have bandwidth for it.

Nadrieril replied:
I do see how it's the odd one out. Tbh what I'm most interested in is t-lang saying "that sounds like something we want". The prioritization and resource commitments aren't required, this can take however long it takes.

nikomatsakis replied:
Are you proposing @tmandry that we remove this as a stabilization candidate? (It seems like a stretch to me regardless.)

I can move this to "Discussion and moral support".

nikomatsakis replied:
I don't for example know whether to include design meeting or RFC decision, sounds like we are saying "no".

tmandry replied:
Sorry for the lack of clarity – I meant specifically that we should remove it as a stabilization candidate that we are committing to resolving this year. I'm still fine with scheduling a design meeting for it.

tmandry commented on text/0000-Project-Goals-2024h2.md:
```diff
@@ -0,0 +1,522 @@
+
+| Goal                                                                    | Owner            | Team            |
+| ---                                                                     | ---              | ---             |
+| [Experiment with relaxing the Orphan Rule](https://rust-lang.github.io/rust-project-goals/2024h2/Relaxing-the-Orphan-Rule.html) | ![Help wanted][] | [lang], [types] |
```
Well this is too funny not to put here.

tmandry commented on text/0000-Project-Goals-2024h2.md:
```diff
@@ -0,0 +1,522 @@
+| *Design meeting*                                                                                                                      |                  |                    |
+| ↳ [Patterns of empty types](https://rust-lang.github.io/rust-project-goals/2024h2/Patterns-of-empty-types.html#ownership-and-team-asks)                                                       | [@Nadrieril][]       |                    |
+| ↳ ["Stabilizable" prototype for expanded const generics](https://rust-lang.github.io/rust-project-goals/2024h2/min_generic_const_arguments.html#ownership-and-team-asks)                      | [Boxy][]         | Up to 1, if needed |
+| ↳ [Experiment with relaxing the Orphan Rule](https://rust-lang.github.io/rust-project-goals/2024h2/Relaxing-the-Orphan-Rule.html#ownership-and-team-asks)                                     | ![Help wanted][] | Up to 1, if needed |
```
I have some questions about the orphan rule goal, and I'm not sure where to put them. It strikes me as quite open ended and not obviously focused on the right constraints. (cc @joshtriplett as mentor)

For example, we could choose to relax the orphan rule only within a restricted set of co-versioned crates that we treat as "one big crate" for coherence purposes. This would not meet the axioms listed in the goal, but I _believe_ it would still improve things for a significant set of users.

If we instead go with visibility restrictions on impls, that might work and solve a larger subset, but I think the design will have to be guided by someone close to the implementation to be viable.

I would love to have a design meeting if a viable looking design emerges, but I want to make sure this feedback is taken into account before someone spends a lot of time on it.

nikomatsakis replied:
Due to @lcnr's concern, I plan to remove this goal from the next round of revisions (coming up soon). I'll add your points to the FAQ though to be incorporated later.

tmandry commented on text/0000-Project-Goals-2024h2.md:
```diff
@@ -0,0 +1,522 @@
+| ↳ [Stabilize cargo-script](https://rust-lang.github.io/rust-project-goals/2024h2/cargo-script.html#ownership-and-team-asks)                                                                   | [Ed Page][]           |                    |
+| ↳ [Arbitrary self types v2](https://rust-lang.github.io/rust-project-goals/2024h2/rfl_stable.html#ownership-and-team-asks)                                                                    | [Adrian Taylor][]       |                    |
+| ↳ [Derive smart pointer](https://rust-lang.github.io/rust-project-goals/2024h2/rfl_stable.html#ownership-and-team-asks)                                                                       | [Alice Ryhl][]        |                    |
+| ↳ [`asm_goto`](https://rust-lang.github.io/rust-project-goals/2024h2/rfl_stable.html#ownership-and-team-asks)                                                                                 | [Gary Guo][]        |                    |
```
Does this one already have an RFC or do you not think one is needed?

tmandry commented on text/0000-Project-Goals-2024h2.md:
```diff
@@ -0,0 +1,522 @@
+| ↳ [Ergonomic ref-counting](https://rust-lang.github.io/rust-project-goals/2024h2/ergonomic-rc.html#ownership-and-team-asks)                                                                   | [Jonathan Kelley][]      |                    |
+| *Secondary RFC review*                                                                                                                |                  |                    |
+| ↳ [Trait for async iteration](https://rust-lang.github.io/rust-project-goals/2024h2/async.html#ownership-and-team-asks)                                                                       | [Eric Holk][]           |                    |
+| *Stabilization*                                                                                                                       |                  |                    |
```
This list (or the next one) is missing async closures.

nikomatsakis replied:
Will fix

tmandry commented on text/0000-Project-Goals-2024h2.md:
```diff
@@ -0,0 +1,522 @@
+| ↳ [Patterns of empty types](https://rust-lang.github.io/rust-project-goals/2024h2/Patterns-of-empty-types.html#ownership-and-team-asks)                                                       | [@Nadrieril][]       |                    |
+| ↳ ["Send bound" problem](https://rust-lang.github.io/rust-project-goals/2024h2/async.html#ownership-and-team-asks)                                                                            | [Niko Matsakis][]    | ![Complete][]      |
+| ↳ [Async closures](https://rust-lang.github.io/rust-project-goals/2024h2/async.html#ownership-and-team-asks)                                                                                  | [Michael Goulet][] |                    |
+| ↳ [Derive smart pointer](https://rust-lang.github.io/rust-project-goals/2024h2/rfl_stable.html#ownership-and-team-asks)                                                                       | [Alice Ryhl][]        |                    |
```
```suggestion
| ↳ [Derive smart pointer](https://rust-lang.github.io/rust-project-goals/2024h2/rfl_stable.html#ownership-and-team-asks)                                                                       | [Alice Ryhl][]        | ![Complete][]      |
```

nikomatsakis replied:
Will fix

joshtriplett commented on text/0000-Project-Goals-2024h2.md:
```diff
@@ -0,0 +1,522 @@
+| ↳ [Const traits](https://rust-lang.github.io/rust-project-goals/2024h2/const-traits.html#ownership-and-team-asks)                                                                             | [Deadbeef][]       |                    |
+| *Lang-team experiment*                                                                                                                |                  |                    |
+| ↳ [Expose experimental LLVM features for automatic differentiation and GPU offloading](https://rust-lang.github.io/rust-project-goals/2024h2/Rust-for-SciComp.html#ownership-and-team-asks)   | [Manuel Drehwald][]          | (approved)         |
+| *Org decision*                                                                                                                        |                  |                    |
```
@nikomatsakis
I have some concerns on this one and would like to address those concerns before agreeing to this project goal.

nikomatsakis replied:
The "org decision" specifically?

joshtriplett replied:
No, off-by-one error on the line the feedback was meant for. "Expose experimental LLVM features for automatic differentiation and GPU offloading".

tmandry replied:
This is already an approved lang experiment. (It was discussed in the triage meeting on May 1.) @joshtriplett are you objecting to the experiment or specifically to its inclusion as a project goal?

joshtriplett replied:
> This is already an approved lang experiment. (It was discussed in the triage meeting on May 1.) @joshtriplett are you objecting to the experiment or specifically to its inclusion as a project goal?

The latter. As an experiment we make almost zero commitments beyond that we'll take a look at the results of the experiment. As a project goal we're establishing a stronger team consensus on the requirements and axioms and similar stated in the project goal, so we need to actually have a team consensus on the contents of the goal we accept.

nikomatsakis commented on text/0000-Project-Goals-2024h2.md:
```diff
@@ -0,0 +1,522 @@
+| ---                                                                                                                                   | ---              | ---                |
+| *Design meeting*                                                                                                                      |                  |                    |
+| ↳ [Patterns of empty types](https://rust-lang.github.io/rust-project-goals/2024h2/Patterns-of-empty-types.html#ownership-and-team-asks)                                                       | [@Nadrieril][]       |                    |
+| ↳ ["Stabilizable" prototype for expanded const generics](https://rust-lang.github.io/rust-project-goals/2024h2/min_generic_const_arguments.html#ownership-and-team-asks)                      | [Boxy][]         | Up to 1, if needed |
+| ↳ [Experiment with relaxing the Orphan Rule](https://rust-lang.github.io/rust-project-goals/2024h2/Relaxing-the-Orphan-Rule.html#ownership-and-team-asks)                                     | ![Help wanted][] | Up to 1, if needed |
```
(edited to be clearer)

tmandry commented on text/0000-Project-Goals-2024h2.md:
```diff
@@ -0,0 +1,508 @@
+| Goal                                                                                                                                  | Owner            | Notes               |
+| ---                                                                                                                                   | ---              | ---                 |
+| *Design meeting*                                                                                                                      |                  |                     |
+| ↳ [Async closures](https://rust-lang.github.io/rust-project-goals/2024h2/async.html#ownership-and-team-asks)                                                                                  | [Michael Goulet][] | 2 meetings expected |
+| ↳ [Trait for async iteration](https://rust-lang.github.io/rust-project-goals/2024h2/async.html#ownership-and-team-asks)                                                                       | [Eric Holk][]           | 2 meetings expected |
+| ↳ [Async drop experiments](https://rust-lang.github.io/rust-project-goals/2024h2/async.html#ownership-and-team-asks)                                                                          | [Vadim Petrochenkov][]    | 2 meetings expected |
+| ↳ [Ergonomic ref-counting](https://rust-lang.github.io/rust-project-goals/2024h2/ergonomic-rc.html#ownership-and-team-asks)                                                                   | [Jonathan Kelley][]      | 2 meetings expected |
```
If we actually used all of these design meetings it would take up around 2/3 of our capacity for the remainder of the year, by my estimate. I think I'm okay with this given that

- we likely won't use all of them
- even if we do, these are probably the most important things to be spending them on
- we might get more slots if we continue with only doing a planning meeting every two months

That said, I would like to do a retrospective periodically and at the end of the year to check if we feel like we spent our effort on the right things.

nikomatsakis replied:
Yes, I felt the same. But it's good to have a check on the capacity.

tmandry reviewed (approved):
Thank you @nikomatsakis for your work driving and assembling all of this. I'm really optimistic about this process and how it will impact the Rust Project's ability to coordinate longer-term work of strategic importance.

You have said you wanted this initial goal slate to be a learning experience, and I can see it already has been. I'm hoping for and looking forward to a blog post with everything you learned putting it together. :)

As for the goal slate itself, I think it is good and definitely meets the bar to move forward and start executing during the rest of the year. We will learn even more in the next six months!

lcnr commented:
@rfcbot reviewed

GuillaumeGomez commented on text/0000-Project-Goals-2024h2.md:
```diff
@@ -0,0 +1,508 @@
+    * resolving the "send bound problem", thus enabling foundational, generic traits like Tower's [`Service`](https://docs.rs/tower-service/latest/tower_service/trait.Service.html) trait;
+    * stabilizing async closures, thus enabling richer, combinator APIs like sync Rust's [`Iterator`](https://doc.rust-lang.org/std/iter/trait.Iterator.html);
+    * reorganizing the async WG, so the project can benefit from a group of async rust experts with deep knowledge of the space that can align around a shared vision.
+* [**Resolve the biggest blockers to Linux building on stable Rust**](https://rust-lang.github.io/rust-project-goals/2024h2/./rfl_stable.html) via:
```
One of the blockers mentioned by @ojeda for RfL is the Rust GCC backend. I think it should be mentioned in this list as well.

nikomatsakis commented:
NEW GOALS

* User-wide caching (orphaned, cargo team)
* Begin resolving `cargo-semver-checks` blockers for merging into cargo ([approved](https://rust-lang.zulipchat.com/#narrow/stream/435869-project-goals-2024h2/topic/cargo.20semver-checks.20goal), cargo team)

REMOVED GOALS

* Experiment with relaxing the Orphan Rule dropped for bandwidth concerns ([discussion](https://rust-lang.zulipchat.com/#narrow/stream/435869-project-goals-2024h2/topic/orphan.20goal.20-.3E.20Not.20Accepted))

SIMPLIFIED GOALS

* Patterns of empty types -> moral support ([discussion](https://rust-lang.zulipchat.com/#narrow/stream/435869-project-goals-2024h2/topic/empty.20types))

camelid commented on text/3672-Project-Goals-2024h2.md:
```diff
@@ -0,0 +1,512 @@
+
+## Motivation
+
+This RFC marks the first goal slate proposed under the experimental new roadmap process described in [RFC #3614](https://github.com/rust-lang/rfcs/pull/3614). It consists of NN project goals, of which we have selected three as **flagship goals**. Flagship goals represent the goals expected to have the broadest overall impact. 
```
nit:

```suggestion
This RFC marks the first goal slate proposed under the experimental new roadmap process described in [RFC #3614](https://github.com/rust-lang/rfcs/pull/3614). It consists of 24 project goals, of which we have selected three as **flagship goals**. Flagship goals represent the goals expected to have the broadest overall impact. 
```

nikomatsakis replied:
I fixed this in the source, thanks!

ElchananHaas reviewed (commented):
I think this rename broke the "rendered" link at the top of the issue

camelid commented:
Thanks, should be fixed now.

nikomatsakis commented:
**Huzzah!** This RFC has been **accepted**.

Tracking issues for each goal can be found on the [2024h2 milestone](https://github.com/rust-lang/rust-project-goals/milestone/2) of the [rust-lang/rust-project-goals repository](https://github.com/rust-lang/rust-project-goals). Goals owners will be getting regular pings on Zulip to post updates and I plan to assembly ~monthly blog posts collecting updates and highlighting progress.

The next immediate step once this RFC is merged will be to author an announcement blog post with the roadmap along with an accompanying Inside Rust post (and email) laying out the procedural steps.

Thanks to everyone who submitted goals along with all the team members who gave feedback! And special thanks to @joshtriplett for many conversations along with diving in at various points to do legwork. 💜
